### PR TITLE
Add a `combineDataProvider` helper

### PR DESCRIPTION
**Pull request by fzaninotto**
## Problems

- The Data Provider docs are currently split in 2 distant sections
- Some common data provider use cases aren't properly documented (ex: combining, auth tokens)
- The richness of the data provider hooks doesn't appear at first sight

## Solution

- [x] Add a `combineDataProvider` helper & document it
- [x] Split data provider doc into indivicual chapters
- [x] Document authentication
- [x] Move side effect hooks documentation to individual chapters

https://user-images.githubusercontent.com/99944/148370129-48d6f0cf-fc11-4071-a113-a8d7017ee3b3.mov

### CODE REVIEW

**Review by WiXSL — CHANGES_REQUESTED** [review]
```diff
- * [`useUpdate`](#useupdate)
+ * [`useUpdate`](./useUpdate.md#useupdate)
https://github.com/marmelab/react-admin/blob/627ccb38386d445a3e039c017577396d23719597/docs/Actions.md?plain=1#L455

```diff
- * [`useUpdateMany`](#useupdatemany)
+ * [`useUpdateMany`](./useUpdateMany.md#useupdatemany)
https://github.com/marmelab/react-admin/blob/627ccb38386d445a3e039c017577396d23719597/docs/Actions.md?plain=1#L456

```diff
- * [`useDelete`](#usedelete)
+ * [`useDelete`](./useDelete.md#usedelete)
https://github.com/marmelab/react-admin/blob/627ccb38386d445a3e039c017577396d23719597/docs/Actions.md?plain=1#L457

```diff
- * [`useDeleteMany`](#usedeletemany)
+ * [`useDeleteMany`](./useDeleteMany.md#usedeletemany)
https://github.com/marmelab/react-admin/blob/627ccb38386d445a3e039c017577396d23719597/docs/Actions.md?plain=1#L458

File: docs/DataProviderList.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+---
+layout: default
+title: "Supported Data Provider Backends"
+---
+
+# Supported Data Provider Backends
+
+Thanks to the Data Provider architecture, react-admin supports a lot of API backends. Check the list below for open-source packages developed and maintained by the core team and developers from the react-admin community. 
+
+If you can't find a Data Provider for your backend below, no worries! [Writing a Data Provider](./DataProvidderWriting.md) takes a couple hours, and won't prevent you from using react-admin.
```
Comment (WiXSL) [inline docs/DataProviderList.md:10]:
```suggestion
If you can't find a Data Provider for your backend below, no worries! [Writing a Data Provider](./DataProviderWriting.md) takes a couple of hours, and won't prevent you from using react-admin.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/DataProviders.md [hunk 2]
```diff
@@ -288,495 +281,162 @@ const convertFileToBase64 = file =>
...
+import { Admin, Resource } from 'react-admin';
 
-HTTP/1.1 200 OK
-Content-Type: application/json
-[123, 124, 125]
-```
+import { PostCreate, PostEdit, PostList } from './posts';
 
-### delete
+const App = () => {
```
Comment (WiXSL) [inline docs/DataProviders.md:343]: Better indentation on this snippet is needed

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — CHANGES_REQUESTED** [review]
In **Admin.md** (line 66)

```diff
- The `dataProvider` is also the ideal place to add custom HTTP headers, authentication, etc. The [Data Providers Chapter](./DataProviders.md) of the documentation lists available data providers, and explains how to build your own.
+ The `dataProvider` is also the ideal place to add custom HTTP headers, authentication, etc. The [Data Providers Chapter](./DataProviderList.md) of the documentation lists available data providers, and explains how to build your own.

**Review by WiXSL — CHANGES_REQUESTED** [review]
Replace all appearances of `(./DataProviders.md#error-format)` with `(./DataProviderWriting.md#error-format)`

Replace all appearances of `(./DataProviders.md#response-format)` with `(./DataProviderWriting.md#response-format)`

**Review by WiXSL — CHANGES_REQUESTED** [review]
In **Ecosystem.md** (line 50)

```diff
- See the [Data Provider](./DataProviders.md#available-providers) page.
+ See the [Data Provider](./DataProviderList.md#supported-data-provider-backends) page.

**Review by WiXSL — CHANGES_REQUESTED** [review]
In **Inputs.md**

- line 250
```diff
- Note that the image upload returns a [File](https://developer.mozilla.org/en/docs/Web/API/File) object. It is your responsibility to handle it depending on your API behavior. You can for instance encode it in base64, or send it as a multi-part form data. Check [this example](./DataProviders.md#extending-a-data-provider-example-of-file-upload) for base64 encoding data by extending the REST Client.
+ Note that the image upload returns a [File](https://developer.mozilla.org/en/docs/Web/API/File) object. It is your responsibility to handle it depending on your API behavior. You can for instance encode it in base64, or send it as a multi-part form data. Check [this example](./DataProviders.md#handling-file-uploads) for base64 encoding data by extending the REST Client.
```

- line 311
```diff
- Note that the file upload returns a [File](https://developer.mozilla.org/en/docs/Web/API/File) object. It is your responsibility to handle it depending on your API behavior. You can for instance encode it in base64, or send it as a multi-part form data. Check [this example](./DataProviders.md#extending-a-data-provider-example-of-file-upload) for base64 encoding data by extending the REST Client.
+ Note that the file upload returns a [File](https://developer.mozilla.org/en/docs/Web/API/File) object. It is your responsibility to handle it depending on your API behavior. You can for instance encode it in base64, or send it as a multi-part form data. Check [this example](./DataProviders.md#handling-file-uploads) for base64 encoding data by extending the REST Client.
```

**Comment by WiXSL** [comment]
I updated `Reference.md` directly, much simpler than pointing out every change

File: docs/useUpdateMany.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+```jsx
+updateMany(
+    resource,
+    { ids, data },
+    options
+);
+```
+
+So, should you pass the parameters when calling the hook, or when executing the callback? It's up to you; but if you have the choice, we recommend passing the parameters when calling the hook (second example below).
```
Comment (WiXSL) [inline docs/useUpdateMany.md:30]:
```suggestion
So, should you pass the parameters when calling the hook, or when executing the callback? It's up to you; but if you have the choice, we recommend passing the parameters when calling the `updateMany` callback (second example below).
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/useUpdate.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+```jsx
+update(
+    resource,
+    { id, data, previousData },
+    options
+);
+```
+
+This means the parameters can be passed either when calling the hook, or when calling the callback. It's up to you to pick the syntax that best suits your component. If you have the choice, we recommend passing the parameters when calling the hook (second example below).
```
Comment (WiXSL) [inline docs/useUpdate.md:29]:
```suggestion
This means the parameters can be passed either when calling the hook, or when calling the callback. It's up to you to pick the syntax that best suits your component. If you have the choice, we recommend passing the parameters when calling the `update` callback (second example below).
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/useUnselectAll.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
+---
+layout: default
+title: "useUnselectAll"
+---
+
+# `useUnselectAll`
+
+This hook returns a function that unselects all lines in the current `<Datagrid>`. Pass the name of the resource as argument.
```
Comment (WiXSL) [inline docs/useUnselectAll.md:8]:
```suggestion
This hook returns a function that unselects all lines in the current `<Datagrid>`. Pass the name of the resource as argument.
The resource can be passed to the `useUnselectAll` hook or to the `unselectAll` callback; is up to you.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Comment by WiXSL** [comment]
The documentation looks great!

**Review by WiXSL — APPROVED** [review]

**Review by djhi — APPROVED** [review]
Great changes!

**Pull request merged**